## 异常处理

### 错误类型

- 语法错误

  代码的拼写错误，也成为编译错误   SyntaxError

- 运行时错误

  在解释执行过程中产生的错误（导包，除数为零，文件不存在）

- 逻辑错误

  程序可以正常运行，但结果不正确


### 异常处理机制

```python
try:
    pass
except FileNotFoundError:     #只能写一个
    pass    #捕获异常
except PermissionError:
    pass
else:
    pass   #没有发生异常时执行的语句
finally:
    pass    #一定会执行
```

```python
try   expect    else    finally
try   except    else
try   except
try   finally
```

```python
BaseException
Exception    SystemExit   KeyboardInterrupt   GeneratorExit
```

```python
Exception
NameError   #尝试放问你一个没有声明的变量
SyntaxError   #语法错误
AttributeError   #尝试访问位置对象的属性
TypeError   #类型错误
ValueError   # 数值错误
ZeroDivisionError   #零除错误
IndexError   #索引超出边界
KeyError   #字典关键字不存在
```

```python
raise   #异常抛出

a = -9
if a<0:
    raise ValueError('数值不能为负数')
```

捕获异常顺序先写子类再写父类



### 自定义异常类

自定义异常类都要继承Exception或其子类

自定义异常类命名一般以Error或Exception为后缀

```python
class NumberError(Exception):
    def __init__(self,data):
        Exception.__init__(self.data)
        self.data = data
    def __str__(self):
        return self.data + '非法数值'
    
def total(data):
    total = 0
    for i in data:
        if i<0:raise NumberError(str(i))
            total += i
    return total 
```



### 断言

一般形式：

- 前置条件断言
- 后置条件断言
- 前后不变断言

```python
assert 布尔类型的表达式|结果是布尔类型的函数  #不建议使用,无提示信息
assert 布尔类型的表达式，字符串表达式
```

------

*start*

Python解释器

两种运行模式：调试模式(启用断言)   优化模式(禁用断言)

通常为调试模式，内置只读变量__ debug __的值为True

但如果使用选项-O运行时，模式为优化模式__ debug __的值False

```python
if __debug__:
    if not testexpression:raise AsssertionError
```

```python
if __debug__:
    if not testexception:raise AssertionError if __debug:
            if not testexception:raise AssertionError(data)
```

*end*

------

```python
a = int(input('A的值'))
b = int(input('B的值'))
assert b!=0,'除数不为0'
c = a/b
print(a,'/',b,'=',c)
```

```python
assert 1+1=2
assert isinstance('hello',str)
assert isinstance('hello',int)
```

```python
s = 'nothing is impossible'
key = 'nothing'
assert key in s,"key:{} is not in Target:{}".format(key,s)
```

改进方式：采用开源的测试框架进行断言操作

- py.test

  轻量级的测试框架

  ```python
  pip install -U pytest
  pytest --version
  ```

  ```python
  import pytest
  def test_case():
      excepted = 'hello'
      actual = 'Hello'
      assert excepted == actual
  if __name__=='__main__':
      pytest.main()
      
  # 运行方式  dos命令
  py.test AssertTest.py     #运行文件名
  ```

- unittest

  python自带的单元测试框架

  使用该测试框架的时候不推荐使用assert语句，需要使用断言方法self.assertXXX()

  ```python
  import unittest
  class TestStringMethod(unittest.TestCase):
      def test_upper(self):
          self.assertEqual('foo'.upper(),'Foo')
  if __name__=="__main__":
      unittest.main()
  ```

- ptest

  ```python
  pip install ptest
  ```

  ```python
  from ptest.decorator import *
  from ptest.assertion import *
  
  @TestClass()
  class TestCases:
      @Test()
      def test(self):
          actual = 'Foo'
          expected = 'bar'
          assert_that(expected).is_equals_to(actual)
          
  #运行
  #python2
  ptest -t 文件名.类名.方法名
  #python3
  ptest3 -t 文件名   # 所有类所有方法
  ptest3 -t 文件名.类名.方法名    # 某一特定方法
  ```

- assertpy

  ```python
  pip install assertpy
  ```

  ```python
  from assertpy import assert_that
  def test_something():
      assert_that(1+2).is_equal_to(33)
      assert_that('foobars').is_length(6).starts_with('foo').ends_with('bar')
      assert_that(['a','b']).contains('a').does_not_contains('x')
  if __name__=='__main__':
      test_something()
  ```